### PR TITLE
[MM-32343] Environment SubSection

### PR DESCRIPTION
**Pull request by maxerenberg**
#### Summary
This PR adds subsections in the system console for ENVIRONMENT permissions.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-32343

### CODE REVIEW

File: components/admin_console/admin_definition.jsx [hunk 1]
```diff
@@ -574,12 +574,29 @@ const AdminDefinition = {
         icon: 'fa-server',
         sectionTitle: t('admin.sidebar.environment'),
         sectionTitleDefault: 'Environment',
-        isHidden: it.not(it.userHasReadPermissionOnResource('environment')),
+        isHidden: it.all(
+            it.not(it.userHasReadPermissionOnResource('environment.web_server')),
```
Comment (BenCookie95) [inline components/admin_console/admin_definition.jsx:578]: I think all these should be using the `RESOURCE_KEYS` constant your redux PR

File: i18n/en.json [hunk 2]
```diff
@@ -1,4 +1,14 @@
...
+  "1": "",
+  "3": "",
+  "": "",
+  "'": "",
+  "\"": "",
+  "@": "",
+  "\\+": "",
+  "=": "",
+  "0.8": "",
+  "1.25": "",
```
Comment (BenCookie95) [inline i18n/en.json:11]: You can remove these

File: i18n/en.json [hunk 3]
```diff
@@ -4054,6 +4084,7 @@
   "unarchive_channel.confirm": "Confirm UNARCHIVE Channel",
   "unarchive_channel.del": "Unarchive",
   "unarchive_channel.viewArchived.question": "Are you sure you wish to unarchive the **{display_name}** channel?",
+  "undefined": "",
```
Comment (BenCookie95) [inline i18n/en.json:4087]: You can remove all these new ones that have no value

**Review by BenCookie95 — CHANGES_REQUESTED** [review]

File: i18n/en.json (hunk 3 above)
Comment (maxerenberg) [reply i18n/en.json:4087]: These appear to be added by the `make i18n-extract` command. Not too sure why that's happening, but I'm going to remove them and push again.

Comment (hahmadia) [reply i18n/en.json]: @maxerenberg See if you have `storybook-static` anywhere. Delete it. It's probably gitignored. I got my PR merged for the thing that does `i18n-extract` but we haven't updated master yet (I'm doing that now as I type).

Here is the PR https://github.com/mattermost/mattermost-utilities/pull/49/files . Just delete `storybook-static` and you should be gucci. I had the same problem for a long time before figuring out whats wrong.

File: components/admin_console/admin_definition.jsx [hunk 4]
```diff
@@ -1091,16 +1118,19 @@ const AdminDefinition = {
...
                         success_message_default: 'Connection was successful',
-                        isDisabled: it.not(it.userHasReadPermissionOnResource('environment')),
+                        isDisabled: it.not(it.userHasReadPermissionOnResource(RESOURCE_KEYS.ENVIRONMENT.FILE_STORAGE)),
                     },
                 ],
             },
         },
         image_proxy: {
-            url: 'environment/image_proxy',
+            url: RESOURCE_KEYS.ENVIRONMENT.IMAGE_PROXY,
```
Comment (hahmadia) [inline components/admin_console/admin_definition.jsx:1127]: lets revert this

File: components/admin_console/admin_definition.jsx [hunk 5]
```diff
@@ -1170,10 +1200,13 @@ const AdminDefinition = {
             },
         },
         smtp: {
-            url: 'environment/smtp',
+            url: RESOURCE_KEYS.ENVIRONMENT.SMTP,
```
Comment (hahmadia) [inline components/admin_console/admin_definition.jsx:1203]: lets revert this

File: components/admin_console/admin_definition.jsx [hunk 6]
```diff
@@ -1309,20 +1342,24 @@ const AdminDefinition = {
                 'admin.email.pushContentTitle',
                 'admin.email.pushContentDesc',
             ],
-            isHidden: it.configIsTrue('ExperimentalSettings', 'RestrictSystemAdmin'),
-            isDisabled: it.not(it.userHasWritePermissionOnResource('environment')),
+            isHidden: it.any(
+                it.configIsTrue('ExperimentalSettings', 'RestrictSystemAdmin'),
+                it.not(it.userHasReadPermissionOnResource(RESOURCE_KEYS.ENVIRONMENT.SMTP)),
+            ),
+            isDisabled: it.not(it.userHasWritePermissionOnResource(RESOURCE_KEYS.ENVIRONMENT.SMTP)),
```
Comment (hahmadia) [inline components/admin_console/admin_definition.jsx:1349]:
```suggestion
                it.not(it.userHasReadPermissionOnResource(RESOURCE_KEYS.ENVIRONMENT.PUSH_NOTIFICATION_SERVER)),
            ),
            isDisabled: it.not(it.userHasWritePermissionOnResource(RESOURCE_KEYS.ENVIRONMENT.PUSH_NOTIFICATION_SERVER)),
```

File: components/admin_console/admin_definition.jsx [hunk 7]
```diff
@@ -1309,20 +1342,24 @@ const AdminDefinition = {
...
             schema: {
                 id: 'PushNotificationsSettings',
                 component: PushNotificationsSettings,
             },
         },
         high_availability: {
-            url: 'environment/high_availability',
+            url: RESOURCE_KEYS.ENVIRONMENT.HIGH_AVAILABILITY,
```
Comment (hahmadia) [inline components/admin_console/admin_definition.jsx:1356]: Lets revert

File: components/admin_console/admin_definition.jsx [hunk 8]
```diff
@@ -1346,17 +1383,20 @@ const AdminDefinition = {
...
-            isDisabled: it.not(it.userHasWritePermissionOnResource('environment')),
+            isDisabled: it.not(it.userHasWritePermissionOnResource(RESOURCE_KEYS.ENVIRONMENT.HIGH_AVAILABILITY)),
             schema: {
                 id: 'ClusterSettings',
                 component: ClusterSettings,
             },
         },
         rate_limiting: {
-            url: 'environment/rate_limiting',
+            url: RESOURCE_KEYS.ENVIRONMENT.RATE_LIMITING,
```
Comment (hahmadia) [inline components/admin_console/admin_definition.jsx:1393]: Lets revert

File: components/admin_console/admin_definition.jsx [hunk 9]
```diff
@@ -1463,10 +1503,13 @@ const AdminDefinition = {
             },
         },
         logging: {
-            url: 'environment/logging',
+            url: RESOURCE_KEYS.ENVIRONMENT.LOGGING,
```
Comment (hahmadia) [inline components/admin_console/admin_definition.jsx:1506]: Lets revert this

File: components/admin_console/admin_definition.jsx [hunk 10]
```diff
@@ -1587,16 +1630,19 @@ const AdminDefinition = {
...
                         },
-                        isDisabled: it.not(it.userHasWritePermissionOnResource('environment')),
+                        isDisabled: it.not(it.userHasWritePermissionOnResource(RESOURCE_KEYS.ENVIRONMENT.LOGGING)),
                     },
                 ],
             },
         },
         session_lengths: {
-            url: 'environment/session_lengths',
+            url: RESOURCE_KEYS.ENVIRONMENT.SESSION_LENGTHS,
```
Comment (hahmadia) [inline components/admin_console/admin_definition.jsx:1639]: Lets revert this

File: components/admin_console/admin_definition.jsx [hunk 11]
```diff
@@ -1616,19 +1662,20 @@ const AdminDefinition = {
...
-            isDisabled: it.not(it.userHasWritePermissionOnResource('environment')),
+            isDisabled: it.not(it.userHasWritePermissionOnResource(RESOURCE_KEYS.ENVIRONMENT.SESSION_LENGTHS)),
             schema: {
                 id: 'SessionLengths',
                 component: SessionLengthSettings,
             },
         },
         metrics: {
-            url: 'environment/performance_monitoring',
+            url: RESOURCE_KEYS.ENVIRONMENT.PERFORMANCE_MONITORING,
```
Comment (hahmadia) [inline components/admin_console/admin_definition.jsx:1672]: Lets revert this

File: components/admin_console/admin_definition.jsx [hunk 12]
```diff
@@ -1654,16 +1701,19 @@ const AdminDefinition = {
...
                         help_text_default: 'The address the server will listen on to expose performance metrics.',
-                        isDisabled: it.not(it.userHasWritePermissionOnResource('environment')),
+                        isDisabled: it.not(it.userHasWritePermissionOnResource(RESOURCE_KEYS.ENVIRONMENT.PERFORMANCE_MONITORING)),
                     },
                 ],
             },
         },
         developer: {
-            url: 'environment/developer',
+            url: RESOURCE_KEYS.ENVIRONMENT.DEVELOPER,
```
Comment (hahmadia) [inline components/admin_console/admin_definition.jsx:1710]: Lets revert this

File: e2e/cypress/support/api/role.js [hunk 13]
```diff
@@ -136,7 +135,32 @@ export const defaultRolesPermissions = {
...
+        'sysconsole_write_environment_file_storage',
+        'sysconsole_write_environment_image_proxy',
+        'sysconsole_write_environment_smtp',
+        'sysconsole_write_environment_push_notification_server',
+        'sysconsole_write_environment_high_availability',
+        'sysconsole_write_environment_rate_limiting',
+        'sysconsole_write_environment_logging',
+        'sysconsole_write_environment_session_lengths',
+        'sysconsole_write_environment_performance_monitoring',
+        'sysconsole_write_environment_developer',
```
Comment (hahmadia) [inline e2e/cypress/support/api/role.js:163]: Make sure you're also giving any ancillary permissions they should be getting with these

File: e2e/cypress/support/api/role.js [hunk 14]
```diff
@@ -214,7 +237,32 @@ export const defaultRolesPermissions = {
...
```
Comment (hahmadia) [inline e2e/cypress/support/api/role.js:265]: Make sure you also give them all the ancillary permissions that come with these

File: e2e/cypress/support/api/role.js [hunk 15]
```diff
@@ -245,7 +293,19 @@ export const defaultRolesPermissions = {
...
+        'sysconsole_read_environment_file_storage',
+        'sysconsole_read_environment_image_proxy',
+        'sysconsole_read_environment_smtp',
+        'sysconsole_read_environment_push_notification_server',
+        'sysconsole_read_environment_high_availability',
+        'sysconsole_read_environment_rate_limiting',
+        'sysconsole_read_environment_logging',
+        'sysconsole_read_environment_session_lengths',
+        'sysconsole_read_environment_performance_monitoring',
+        'sysconsole_read_environment_developer',
```
Comment (hahmadia) [inline e2e/cypress/support/api/role.js:308]: Make sure they get any ancillary permissions too!

**Review by hahmadia — CHANGES_REQUESTED** [review]
Nice work! I really like what I see 😃 Just some changes that need to be made.

File: i18n/en.json (hunk 3 above)
Comment (hahmadia) [reply i18n/en.json]: Master just got updated with the latest hash for mm utilities (https://github.com/mattermost/mattermost-webapp/pull/7639). Pull latest master and see if this still happens

Comment (maxerenberg) [reply i18n/en.json]: Thanks, that worked!

**Review by sbishel — COMMENTED** [review]
Ping after updates for Hossein comments.

**Review by sbishel — APPROVED** [review]

File: e2e/cypress/support/api/role.js (hunk 14 above)
Comment (maxerenberg) [reply e2e/cypress/support/api/role.js:277]: Done.

File: e2e/cypress/support/api/role.js (hunk 15 above)
Comment (maxerenberg) [reply e2e/cypress/support/api/role.js:332]: Done.

**Comment by maxerenberg** [comment]
> A small PS. Don't forget to update package.json and package-lock.json mattermost-redux hash with the latest commit in your redux PR so QA will not run into problems. Currently the latest hash is `c8aca93dfaaa81a5c13eb25659f2531db9af42ff`. Thanks

Didn't know that. Thanks for letting me know. :)

**Review by BenCookie95 — APPROVED** [review]
LGTM, nice job!

**Comment by sbishel** [comment]
Performed QA on this set of PRs. Test write, read and no access for each subsection and ability to perform ancillary actions.

**Pull request merged**